practicas/FredinVazquez/P3/P3_Language_Models_Word_Vectors.py

```diff
@@ -0,0 +1,447 @@
+# - Las probabilidades conjuntas
+# $$p(w_i,w_j) = \frac{c_{i,j}}{\sum_i \sum_j c_{i,j}}$$
+# - Las probabilidades marginales
+# $$p(w_i) = \sum_j p(w_i,w_j)$$
+# - Positive Point Wise Mutual Information (PPMI):
+# $$PPMI(w_i,w_j) = \max\{0, \log_2 \frac{p(w_i,w_j)}{p(w_i)p(w_j)}\}$$
+# 
+# **Comparación de representaciones**
+# 
+# Aplica reducción de dimensionalidad de los vectores de la matríz con PPMI y de los vectores entrenados en español:
+# 
+# - Realiza un plot de 100 vectores aleatorios (que esten tanto en la matríz como en los vectores entrenados)
+# - Compara los resultados de los plots:
+#     - ¿Qué representación dirías que captura mejor relaciones semánticas?
+#     - Realiza un cuadro comparativo de ambos métodos con ventajas/desventajas
+# 
+# ### 📁 [Carpeta con vectores](https://drive.google.com/drive/folders/1reor2FGsfOB6m3AvfCE16NOHltAFjuvz?usp=drive_link)
+
+# ---
+# # Desarrollo
+
+# Obtenga la matriz de co-ocurrencia para un corpus en español y realice los siguientes calculos:
+# - Las probabilidades conjuntas
+# $$p(w_i,w_j) = \frac{c_{i,j}}{\sum_i \sum_j c_{i,j}}$$
+# - Las probabilidades marginales
+# $$p(w_i) = \sum_j p(w_i,w_j)$$
+# - Positive Point Wise Mutual Information (PPMI):
+# $$PPMI(w_i,w_j) = \max\{0, \log_2 \frac{p(w_i,w_j)}{p(w_i)p(w_j)}\}$$
+# 
+# ---
+# 
+# 
+# Para la matriz de co-ocurrencia se hace uso de la función creada en el notebook de la ayudantía.
+
+# In[98]:
+
+
+from datasets import load_dataset
+import pandas as pd
+import numpy as np
+
+import nltk
+from collections import Counter
+from itertools import combinations
+from nltk.corpus import stopwords
+import re
+
+nltk.download('stopwords')
+stop_words = set(stopwords.words('spanish'))
+
+
+# In[99]:
+
+
+ds = load_dataset("mariagrandury/fake_news_corpus_spanish")
+
+# Otros datasets:
+#ds = load_dataset("spanish-ir/messirve", "ar")
+
+
+# In[ ]:
+
+
+##############################
+# Cargando un Corpus en español
+##############################
+
+# Preprocesando texto
+def preprocesando_corpus(documentos):
+  """
+    Preprocesa una lista de documentos aplicando limpieza, eliminación de stopwords
+    y filtrado de palabras de longitud menor o igual a 1.
+
+    Parameters
+    ----------
+    documentos : list of str
+        Una lista de cadenas donde cada elemento representa un documento de texto
+        que será sometido a preprocesamiento.
+
+    Returns
+    -------
+    list of list of str
+        Una lista donde cada elemento es una lista de palabras que representa el
+        documento procesado.
+
+  """
+  
+  documentos_preprocesados = []
+
+  for documento in documentos:
+      documento = documento.lower()
+
+      documento = re.sub(r'[^a-záéíóúñü\s]', '', documento)
+
+      palabras = documento.split()
+
+      palabras_filtradas = [palabra for palabra in palabras if palabra not in stop_words]
+
+      palabras_filtradas = [palabra for palabra in palabras if len(palabra)>1]
```

Comment:
Quizá esto pudo estar en una sola línea

```suggestion
      palabras_filtradas = [palabra for palabra in palabras if palabra not in stop_words and len(palabra) > 1]
```